[DTensor] fix copy_ strategy

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #158538
* #158490

The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @d4l3k @pragupta

wconstab committed:
[DTensor] fix copy_ strategy

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] fix copy_ strategy"




cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] fix copy_ strategy"




cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] fix copy_ strategy"


The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)

These are the correct sharding combinations:

   self   |     src
------------------------
Shard(0)  |   Replicate()
Shard(1)  |   Replicate()
Shard(2)  |   Shard(0)
Shard(3)  |   Shard(1)

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] fix copy_ strategy"


The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

[ghstack-poisoned]

wconstab commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -93,44 +93,54 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
-            for strategy in first_input_strategy.strategies
-        ]
-    )
+        )
```
oh yea! thanks. i had it in my first version and then forgot about it.

Can we actually make it a required argument for creating an OpSpec?

zpcore reviewed (approved):
LGTM!

XilunWu commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
+    assert isinstance(op_schema.args_schema[0], OpStrategy)
+    assert isinstance(op_schema.args_schema[1], OpStrategy)
+    self_strategy: OpStrategy = op_schema.args_schema[0]
+    mesh = self_strategy.mesh
```
TODO: consider supporting cross-mesh copy

XilunWu reviewed (approved):
some nits

wconstab commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
-    #         self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+    @with_comms
+    def test_copy_broadcast_redistribute(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
```
i am just sticking with the convention in the file. I also want to change things to MultiProcContinuousTest.  (or  threaded, via PR #158082, if that can be made to work - so ill leave it alone in this PR

wconstab commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
+        for dst_spec, src_spec in zip(dst_specs, src_specs):
+            src_dtensor = distribute_tensor(src_tensor, device_mesh, src_spec)
+            dst_dtensor = distribute_tensor(dst_tensor, device_mesh, dst_spec)
+            dst_dtensor.copy_(src_dtensor)
+            dst_tensor.copy_(src_tensor)
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
```
actually, i should check the comm counts, i was just lazy.  I should ensure that a redistribute happend.

wanchaol commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
-    # that is invalid for dst tensor.
-    # It is also problematic to assume that shard(0) on src maps to shard(0) on self, since we
-    # may broadcast a new dim to the left or right of 0 when copying.
+def copy_inplace_strategy(op_schema: OpSchema) -> StrategyType:
```
Hmmm I think you probably want to remove the copy strategy, just register it as a pointwise strategy.

wconstab replied:
hmm. i would be ok with that, but does pointwise strategy properly enforce the extra requirement of inplace_ ops, where broadcasting 'self' to match 'src' is NOT allowed?

wanchaol replied:
Yeah, for inplace ops it just follow the first argument

wconstab replied:
ok, yea i think we can try pointwise. Wish I had known about it!

wanchaol commented on torch/distributed/tensor/_ops/_tensor_ops.py:
```diff
@@ -93,44 +93,59 @@ def propagate_single_input_strategy(op_schema: OpSchema) -> StrategyType:
-    assert isinstance(first_input_strategy, OpStrategy)
-    return OpStrategy(
-        [
+    # DTensor semantics for inplace ops also dictates that we may NOT redistribute our 'self' input.
```
I don't quite like the fact that we are implementing broadcasting semantic separately, it would be better if you could just reuse the pointwise strategy for `copy_`

wanchaol reviewed (commented):
Please see inlined comments, I think we should just use register copy_as a pointwise strategy in `_tensor_ops.py`

wconstab committed:
Update on "[DTensor] fix copy_ strategy"


The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
+    def test_copy_broadcast_redistribute(self):
+        device_mesh = DeviceMesh(self.device_type, list(range(self.world_size)))
+        # The src specs in this case are designed to not be compatible with the dst_specs, redistribute should happen
+        src_specs = [[Shard(1)], [Shard(1)], [Shard(1)]]
```
i don't understand? seems ok to me.

wconstab commented on test/distributed/tensor/test_tensor_ops.py:
```diff
@@ -68,22 +68,37 @@ def test_copy_(self):
+            self.assertEqual(dst_dtensor.full_tensor(), dst_tensor)
+
+    @with_comms
+    def test_copy_broadcast(self):
```
note- i prefer separate tests (or using parameterize) so its easy to run a subtest without commenting stuff out.

but its horribly slow with the current multiproc test case, so i merged them for now

wconstab committed:
Update on "[DTensor] fix copy_ strategy"


The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k

[ghstack-poisoned]

wconstab committed:
Update on "[DTensor] fix copy_ strategy"


The previous strategy directly used 'self' input strategy for 'src'
input.  The fixed strategy correctly maps the self dim to src dim
so that it works even if the src input is broadcast.

E.g. for this program, broadcasting will occur on dims 0,1,3 of self.

```
self = torch.ones((2,3,4,5))
src = torch.ones((4,1))
self.copy_(src)
```

These are the correct sharding combinations:

|   self   |     src |
|-------|------|
| Shard(0)  |   Replicate() |
| Shard(1)  |   Replicate() |
| Shard(2)  |   Shard(0) |
| Shard(3)  |   Shard(1) |

cc H-Huang awgu wanchaol fegin fduwjj wz337 d4l3k pragupta

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge -i
